torchqdynamics/mesolve.py

```diff
@@ -1,63 +1,102 @@
-    return -1j * (H @ rho - rho @ H) + _dissipator(Ls, Lsdag, rho).sum(0)
-
-
-def _mesolve_torchdiffeq(
-    H: torch.Tensor,
-    Ls: List[torch.Tensor],
-    rho0: torch.Tensor,
-    tlist: np.ndarray,
-    options: Dict[str, Any],
-) -> torch.Tensor:
-    Ls = torch.stack(Ls)
-    # [speedup] by precomputing the jump operators adjoint
-    Lsdag = Ls.adjoint().resolve_conj()
-
-    def f(t, x):
-        rho = torch.view_as_complex(x)
-        drho = _lindbladian(H, Ls, Lsdag, rho)
-        return torch.view_as_real(drho)
-
-    y0 = torch.view_as_real(rho0)
-    t = torch.from_numpy(tlist)
-    options = {'step_size': options['step_size']}
-
-    ys = torchdiffeq.odeint(f, y0, t, method='rk4', options=options)
-    return torch.view_as_complex(ys)
+    H, jump_ops, rho0, tsave, solver=Rouchon(), sensitivity='autograd',
+    model_params=None
+):
+    # Define the QSolver
+    if isinstance(solver, Rouchon):
+        if solver.order == 1:
+            qsolver = MERouchon1(H, jump_ops, solver)
+        elif solver.order == 2:
+            qsolver = MERouchon2(H, jump_ops, solver)
+    else:
+        raise NotImplementedError
+
+    # Compute the result
+    return odeint(
+        qsolver, rho0, tsave, sensitivity=sensitivity, model_params=model_params
+    )
+
+
+# --------------------------------------------------------------------------------------
+#     ME QSolver classes
+# --------------------------------------------------------------------------------------
+
+
+class QSolver:
+    def __init__(self):
+        pass
+
+
+class MERouchon(QSolver):
+    def __init__(self, H, jump_ops, solver_options):
+        self.H = H
+        self.jump_ops = jump_ops
+        self.jumpdag_ops = jump_ops.adjoint()
+        self.sum_nojump = (self.jumpdag_ops @ self.jump_ops).sum(dim=0)
+        self.I = torch.eye(H(0).shape[-1]).to(H(0))
+        self.options = solver_options
+
+
+class MERouchon1(MERouchon):
+    def forward(self, t, dt, rho):
+        """Compute rho(t+dt) using a Rouchon method of order 1"""
+        # Non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # Build time-dependent Kraus operator
+        M0 = self.I - 1j * dt * H_nh
+
+        # Compute rho(t+dt)
+        drho = M0 @ rho @ M0.adjoint()
+        drho += dt * (self.jump_ops @ rho.unsqueeze(0) @ self.jumpdag_ops).sum(dim=0)
+        return drho
+
+    def forward_adjoint(self, t, dt, phi):
+        raise NotImplementedError
+
+
+class MERouchon2(MERouchon):
+    def forward(self, t, dt, rho):
+        """Compute rho(t+dt) using a Rouchon method of order 2"""
+        # Non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # Build time-dependent Kraus operators
+        # TODO: Add the missing time derivative term in -0.5j * dt**2 * \dot{H}
+        M0 = self.I - 1j * dt * H_nh - 0.5 * dt**2 * H_nh @ H_nh
+        M1s = 0.5 * self.Ls @ M0
+        M1s += M1s.adjoint()
```

Comment:
(NB: Bug here needs to be fixed. Line 79 should be in line 78 without the adjoint)